plugins/single_plugins/ipc-rules.cpp

```diff
@@ -524,33 +628,62 @@ class ipc_rules_t : public wf::plugin_interface_t, public wf::per_output_tracker
     wf::signal::connection_t<wf::output_plugin_activated_changed_signal> on_plugin_activation_changed =
         [=] (wf::output_plugin_activated_changed_signal *ev)
     {
+        if (clients.empty())
+        {
+            return;
+        }
+
         nlohmann::json data;
         data["event"]  = "plugin-activation-state-changed";
         data["plugin"] = ev->plugin_name;
         data["state"]  = ev->activated;
-        data["output"] = ev->output ? (int)ev->output->get_id() : -1;
+        data["output"] = output_to_json(ev->output);
+        send_event_to_subscribes(data, data["event"]);
+    };
+
+    wf::signal::connection_t<wf::output_gain_focus_signal> on_output_gain_focus =
+        [=] (wf::output_gain_focus_signal *ev)
+    {
+        if (clients.empty())
+        {
+            return;
+        }
+
+        nlohmann::json data;
+        data["event"]  = "output-gain-focus";
+        data["output"] = output_to_json(ev->output);
         send_event_to_subscribes(data, data["event"]);
     };
 
     wf::signal::connection_t<wf::workspace_set_changed_signal> on_wset_changed =
         [=] (wf::workspace_set_changed_signal *ev)
     {
+        if (clients.empty())
+        {
+            return;
+        }
+
         nlohmann::json data;
         data["event"]    = "output-wset-changed";
-        data["new-wset"] = ev->new_wset ? (int)ev->new_wset->get_id() : -1;
-        data["output"]   = ev->output ? (int)ev->output->get_id() : -1;
+        data["new-wset"] = wset_to_json(ev->new_wset.get());
```

Comment:
This is a breaking change and I'd like to avoid making breaking changes, so if we do this, then let's add additional fields for any new information.

Also, this will cause a crash if any of the wsets or outputs in those signals is NULL (this can happen at a few places) - so you'd need to either guard against that or extend the output_to_json and wset_to_json functions to handle NULL outputs.